Parallel sort based on Morsel-Driven Parallelism paper 4.5 section

kingfolk opened the pull request:
## parallel sort time complex analysis

the major time spending is on localSort and merge phase

### localSort analysis

splitCnt = k, every split contains `n/k` elements need to sort. splitCnt means there are `splitCnt` runs of localSort.

big O of all localSort = (n/k) * log(n/k) * k = n * log(n/k), more splits, less localSort altogehter workload.

### merge analysis

splitCnt = k, every split contains `n/k` elements will be merged in merge phase. assume we have `s` seperators, every split will have `s+1` segment. `n/k/s` represents datasize of each split's merge unit range that is bounded by seperator and will merge into output.

```
big O of all merge = (n/k/s) * k * log(k) * s → s merge work count
                         ↓            ↓
           every segment size        minheap of every split's minimum element
                   = n * log(k) → means every element will push/pop to minheap.
```

### global seperator calc

For each calculated global seperator, a local mapping to each split needs to calculate. This a calculatation is a binary search with complexity `log(n/k)`. And there are `k` splits and `s` seperators to calculate. It sum up to `log(n/k) * k * s`. 

it make sense to keep small number of seperator count and split count, otherwise it will consume lots of time.

### conclusion

combine two phase:

```
O = n * log(n/k) + n * log(k) = n * (log(n/k) + log(k)) = n * log(n)
```
so it actually does not matter for what value seperator count are. whether there are lots of seperators or only few, the time complexity is not changed.

However the more seperators are, the global seperator picking will consume more time. Because more seperators, larger input seperator vector to pick. For better performance, it may be better to keep seperator count a relative small number, like under 100.

## case study for `tpch/16.sql`

Current CT for `tpch/16.sql`

![image](https://github.com/user-attachments/assets/96af2f8f-eda3-4746-bf53-ce9b0340afac)

Previous CT for `tpch/16.sql`

![WeChat870b165df603520d331e8c5d5f2d5003](https://github.com/user-attachments/assets/033855cb-ed45-4f57-ad7c-35dea4914912)

it seems localSort, seperatorSearch, sortMerge task are well parallelized.

DO NOT MERGE UNTIL https://github.com/lingo-db/lingo-db/pull/87 is merged. Current sort task design is depend on https://github.com/lingo-db/lingo-db/pull/87. For the reason `reserveWork`, `consumeWork` from https://github.com/lingo-db/lingo-db/pull/87  are likely merge into master. Parallel Sort implementation is only in `GrowingBuffer.cpp` file. 


## benchmark sql case

The following query can be used to test if sort is optimal because almost all of its computation is spent on sorting.
```
select sum(l_quantity) from (
    select
        l_shipdate,
        l_quantity
    from lineitem order by l_shipdate
) as tb;
```

FIX #86

jungmair commented:
General hint: for this pull request, it would make sense to do some dedicated benchmarking for sorting.
Either by building a standalone benchmark binary or by using queries that heavily rely on sorting large data (e.g., window query on lineitem etc)

jungmair commented:
It would be great, if you could rebase this PR onto main (now including the improvements to the scheduler)
For the sorting tasks, you probably want to use the new `TaskWithImplicitContext` that will set the execution context correctly before each time work is performed

kingfolk commented:
> It would be great, if you could rebase this PR onto main (now including the improvements to the scheduler) For the sorting tasks, you probably want to use the new `TaskWithImplicitContext` that will set the execution context correctly before each time work is performed

Got it

kingfolk commented:
> General hint: for this pull request, it would make sense to do some dedicated benchmarking for sorting. Either by building a standalone benchmark binary or by using queries that heavily rely on sorting large data (e.g., window query on lineitem etc)

## bench case

so i bench tested a quit stupid query for sort

```
select sum(l_quantity) from (
    select
        l_shipdate,
        l_quantity
    from lineitem order by l_shipdate
) as tb;
```

this query generate very pure sorting ct as following:

![WeChate489b3a017e5cddcba51006946860007](https://github.com/user-attachments/assets/275f2ad1-3785-4f80-ae96-a00142619b2d)


this query main computational cost is almost purely on sort operator. i would like to choose this simple query as the one to testing sorting parallism. And as i tested it out on my machine with 8 vcore
- branch parallel-sort: ~330ms
- branch main: ~1850ms
it's about `>5.5x` speed boost, a little below `8x`. it's understandable because there is still some synchronization phases that are not parallelized.

other queries like following always have computational costs other than sort. i can get 3x~4x speed boost.
```
-- the following two queries have the same size result set of lineitem rows,
-- which make it time consuming to write back result to client.
select
    l_shipdate,
    l_quantity
 from lineitem order by l_shipdate;

select
    l_shipmode,
    l_shipdate,
    l_quantity,
    min(l_quantity) OVER (PARTITION BY l_shipmode ORDER BY l_shipdate) AS a
 from lineitem;
```

And the `order by` in inner query of this simple case seems not likely to be optimized in lingodb?(not sure) I do `explain` on mysql and postgresql. it seems mysql does optimization and erase `order by`, but postgresql still keep it.

```
-- postgresql create table and explain example
> CREATE TABLE EMPLOYEE (
  empId INTEGER,
  name TEXT NOT NULL,
  dept TEXT NOT NULL
);

> explain select sum(empId) from (
    select
        empId
    from EMPLOYEE order by empId
) as tb;

                              QUERY PLAN                               
-----------------------------------------------------------------------
 Aggregate  (cost=72.61..72.62 rows=1 width=8)
   ->  Sort  (cost=59.86..61.98 rows=850 width=4)
         Sort Key: employee.empid
         ->  Seq Scan on employee  (cost=0.00..18.50 rows=850 width=4)
(4 rows)
```

## question about current benchmark

so i am aware there is a `benchmark-tpch.py` script. there are few question about the benchmark overall and bench for sorting.

for overall benchmark running, it should be a stable benchmark machine that is run for benchmark for every release and commit check.

because i don't have access to the benchmark machine and "official benchmark" for every edition and commit, i only can sense if i did right from my local machine's bench and ct.

Q1. what is the best way for me to monitor benchmark over time? should i keep a record of my local machine's benchmark score from "before fiber scheduler" to newest edition.

Q2. how should i add this sort benchmark case if i need to add it to project. i will try add a new folder under `data/sql/` dir. if the way i add is causing messy and not the best way, please let me know.

jungmair commented on resources/sql/bench_misc/1.sql:
```diff
@@ -0,0 +1,6 @@
+select sum(l_quantity) from (
```
Please remove the benchmark query from the repository/Makefile again

kingfolk replied:
got it

jungmair commented on src/scheduler/Scheduler.cpp:
```diff
@@ -191,16 +191,16 @@ class Scheduler {
       }
+      task->prev = nullptr;
+      task->next = nullptr;
    }
```
What is this change about?
Is it a bug fix independent of the parallel sort?

kingfolk replied:
all right. removing them cause CI fail. i think `task->next = nullptr;` definitely need to keep, see following,

```
            dequeueTaskLocked(toCoolDown);
            std::lock_guard<std::mutex> lock1(coolingDownMutex);
            if (coolingDownTail) {
               coolingDownTail->next = toCoolDown;
               toCoolDown->prev = coolingDownTail;
               coolingDownTail = toCoolDown;
            }
```

after `dequeueTaskLocked`, only `toCoolDown->prev` is link to coolingDown queue, but `toCoolDown->next` is still linking to task queue's next task.

so i guess it is a good habit to keep following

```
      task->prev = nullptr;
      task->next = nullptr;
```
inside dequeueTaskLocked to reset dequeued element's prev and next link, so as to keep everything clean, no garbage pointer attached which can cause issue in other logic.

jungmair replied:
thanks for validating!
Can you make this change a seperate commit with that description in the commit message?

kingfolk replied:
new fix PR is here https://github.com/lingo-db/lingo-db/pull/92

jungmair commented on src/runtime/GrowingBuffer.cpp:
```diff
@@ -1,14 +1,277 @@
+      uint8_t* datum;
+      size_t srcIdx;
+   };
+   typedef struct HeapDatum HeapDatum;
```
This is unnecessary in C++, right?

jungmair commented on src/runtime/GrowingBuffer.cpp:
```diff
@@ -1,14 +1,277 @@
+   // but we keep two vector for case of debug reason.
+   std::vector<size_t> globalSeperators;
+};
+typedef struct SortSplitState SortSplitState;
```
remove

kingfolk replied:
got it

jungmair commented on src/runtime/GrowingBuffer.cpp:
```diff
@@ -1,14 +1,277 @@
+};
+typedef struct SortContext SortContext;
+
+class SortCopyTask : public lingodb::scheduler::TaskWithImplicitContext {
```
Add a comment on top of this class explaining the "job" of this task

jungmair commented on src/runtime/GrowingBuffer.cpp:
```diff
@@ -1,14 +1,277 @@
+};
+
+
+class SortLocalTask : public lingodb::scheduler::TaskWithImplicitContext {
```
Again: add a comment on top that describes this task a bit more. Especially since it not only does local sorting, but also computes local separators

jungmair commented on src/runtime/GrowingBuffer.cpp:
```diff
@@ -1,14 +1,277 @@
+      }
+      for (size_t i = 0; i < splitCnt; i ++) {
+         std::vector<size_t> sepVec;
+         sctx.localStates.push_back(SortSplitState(0,0,sepVec));
```
If you add default values for the integers, you could just use resize localStates, and avoid the loop

jungmair commented on src/runtime/GrowingBuffer.cpp:
```diff
@@ -37,6 +300,52 @@ class GroupAllocator : public lingodb::runtime::GrowingBufferAllocator {
+      seperatorCnt = workerNum;
+      normalizeSepCnt();
+      return;
+   } else if (sizePerWorker > 131072) {
```
Have you validated that this makes sense?
On my machine (16 cores, 32 with simultaneous multi threading), your benchmark query and TPC-H SF=10, I get this trace. It seams that local sorting is one of the faster parts... So maybe it might sense to allow for larger sizes locally, if it then avoids later work... 

![image](https://github.com/user-attachments/assets/a3c4d348-b49e-4fd9-b0fa-d02a83bbdcbd)

kingfolk replied:
all right. i will tune size number to see what works best.

i make the max splitSize 32768 fixed because of a basic vector computation heuristics that intermediate-size vector computation has better performance due to data locality(whole vector can be loaded into high speed cache). but not check the merge and seperator sync phases performan bad.

kingfolk replied:
although i make a small change to splitCnt logic. but i think this performance tune is not done yet.

for example of SF10, lineitem contains 60million rows. let's say we have small splitCnt which is 64, which will cause about 1million rows per split, this huge size split vector will slow down local sorting and binary search in global seperator search phase maybe due to cache miss.

on my 8 vcore machine, `splitSize = 32768*2` perfrom best for SF10 lineitem sorting. on your 32 thread machine, i guess `splitSize = 32768*8` may be the best performance param. the best splitSize/splitCnt value may be different for different machine.

i also update `global seperator search` complexity analysis on PR's description as follwoing
```
O = log(n/k) * k * s
```
where n is total input size, k is split count, s is seperator count.

The following maybe marked as TODO
1. we should keep a small number of seperator count. i would like to make `seperatorCnt = wokerNum`. and change the way merge split from minheap to mergesort way of merge, which has more degree of parallelism. it may still have chance of not load balance. but it worth to see the result because from `splitCnt = workerNum * 4` to `splitCnt = workerNum` will divide seperator search time by 4.
2. for local sort for huge local array, it worthy to explore other option than `std::sort`. small splitCnt in theory should improve the perfromance. but in reality it drag down the performance maybe because the cache miss.

jungmair replied:
The current "minheap" merge method is fine, leave it like that :)

jungmair reviewed (commented):
You did really great work with this pull request!
I also ran benchmarks on my workstation and performance looks great!

General Feedback:
1. Add a few more comments on that explain the big picture for each step/task
2. Add a comment explaining how you compute the global separators from the local ones (that is not immediatly clear when reading the code)
3. Could you maybe move all sorting-related code to an extra file (e.g., Sorting.cpp)
4. Add your benchmark query to the description of the pull request, and remove it from the repository

jungmair commented:
Regarding the benchmarking: At the moment there is no official benchmarking machine.
I usually just compare against e.g., main on the same machine. 
For performance-sensitive pull-requests I try to validate the performance results myself before merging (at least at the moment)

kingfolk reviewed (commented):
thanks for you review!

it's good to know how benchmark is properly measured now. at my last company, we have a crazy server of 48cores plus lots of memory. i guess it also make sense you local machine works as benchmark machine since you actually solo develop and maintain the whole project.

it's still good to know the benchmark you run and me run shouldn't be lots of different, altought mine should be a lot slow. for that 16 core/32 threads number you mentioned, i could give better value for performance-related parameter.

kingfolk commented:
updated according to the review comment.

for standalone `Sorting.cpp`, i am not sure i did it right. because i am aware in most case `#include "xxx.cpp"` is not encouraged.

i also updated PR's description. add extra time complexity analysis for global seperator search phase. please also check extra comment, especially https://github.com/lingo-db/lingo-db/pull/90#discussion_r2000120339

jungmair commented on src/runtime/GrowingBuffer.cpp:
```diff
@@ -1,14 +1,15 @@
 #include "lingodb/runtime/helpers.h"
 #include "lingodb/utility/Tracer.h"
+
+#include "Sorting.cpp"
```
It would be better, if you would be better if you could create a seperate Sorting.h containing a method that can take a `runtime::FlexibleBuffer` by reference, sorts it and then produces a continuous buffer

jungmair commented on src/runtime/Sorting.cpp:
```diff
@@ -0,0 +1,388 @@
+   std::vector<SortSplitState>& localStates;
+};
+
+class SortCopyTask : public lingodb::scheduler::TaskWithImplicitContext {
```
still missing comment :)

jungmair commented on src/runtime/Sorting.cpp:
```diff
@@ -0,0 +1,388 @@
+// ... | 3, 5, 8, 13, 16, 17, 19, 26, 29, 31, 33, 38, 39 | ...
+//             |              |               |
+//            sep1           sep2            sep3
+// Local seperators are equally distributed throught split.
```
perfect :)

jungmair commented:
In general, the PR looks much better now.
But I rerun the benchmark query and it is now 3x slower...
So maybe you revert the parameters you adapted...

kingfolk commented:
> In general, the PR looks much better now. But I rerun the benchmark query and it is now 3x slower... So maybe you revert the parameters you adapted...

reverted. so i say it maybe still some work not done yet for sort. in theory, big splitSize and few splits decrease time complexity. but in reality, time for std::sort for every split is not strictly accroding the time complexity equation i gave, i blame to cache miss.

also it worthy to try splitSize value of `32768 * 2`  on your machine. see if perform better. the best size differ for different machine i guess.
```
} else if (sizePerWorker > 32768 * 4) {
      // a fixed relative small number should help local sorter operate sort algorithm entirely on cache,
      // resulting in good performance.
-     size_t splitSize = 32768;
+    size_t splitSize = 32768 * 2;
```

kingfolk commented:
fixed SeperatorSearch phase abnormal long running time. please try run benchmark again.

i made a mistake of creating new vector from input like following, this segment vector can be very long and eat lots of memory and cpu time. this is my bad for golang developer that always deal with slice. i thought this way of using `std::vector` is same as golang slice and will not make copy.
```
         auto segment = std::vector(input.begin() + localState.begin, input.begin() + localState.end);
         auto it = std::lower_bound(segment.begin(), segment.end(), globalSepVal, sctx.compareFn);
```
